# Ejercicio 10: Cifrado de Hill

## Criptograma:

ZKEKS HOTSE QFYFG NIULT VMGIT ATPKY TWVOQ ZGDWP OGIEL OMWWJ WTYFK ZYSGM KLEUL BQJGM TWOXT AGSBW OIKXZ SAWQZ KVIJN QAUJP OWTWI SZKRB QYGQV MLOGM NSGIJ PQIGB ZKGNE LQSAG KLQWA NYSMW GNADM QTIIQ VMMOV ORCQM ANRGY RADEK YRGXG EIETK JHYFK ZYSGM INRWT QTWOW TIEUL BKZGO DSART VGXAN ELVOE FSITB OWGEZ GZKMS GNARK VWCNQ TXWRS EGXZS AWQSS GXDGD ESOAG MVWKC OGQYS IADVS NSGVA NGNYB KLOKZ CNQNY SL

## Hints:

+ La 5ta letra es una **e** (pista dada por la profesora)
+ La 1ra letra es una **r** (pista dada por la profesora)
+ La penúltima letra es una **s** (pista dada por el ayudante Luis Ángel)
+ La ultima letra es una **z** (pista dada por el ayudante Luis Ángel)

El criptograma está dividido por bloques, sabemos que en el cifrado de hill se usan matrices cuadradas de nxn para cifrar, por lo que el texto se tiene que dividir en bloques de longitud n. Se tienen bloques de 5 letras y el último bloque tiene solamente 2, por lo que no se pudo usar matrices de 5x5 para cifrar, lo más probable es que se hayan utilizado una matriz de 2x2, puesto que se tienen en total 322 letras. Cabe la posibilidad de que se hayan ocupado matrices de otro tamaño, como por ejemplo matrices de 7x7, sin embargo es poco probable ya que sería un ejercicio muy dificil de descifrar con tan poca información para una tarea, por lo que podemos suponer que se utilizó una matriz de 2x2. Importamos los modulos necesarios para resolver el criptograma y cargamos el criptograma sin espacios.



In [1]:
from cifrar_hill import *

criptograma = None
with open('./archivos/Criptograma_5.txt', 'r') as file:
    # Eliminamos espacios y lo guardamos en una variable
    criptograma = file.read().replace(' ','')
print(criptograma)

ZKEKSHOTSEQFYFGNIULTVMGITATPKYTWVOQZGDWPOGIELOMWWJWTYFKZYSGMKLEULBQJGMTWOXTAGSBWOIKXZSAWQZKVIJNQAUJPOWTWISZKRBQYGQVMLOGMNSGIJPQIGBZKGNELQSAGKLQWANYSMWGNADMQTIIQVMMOVORCQMANRGYRADEKYRGXGEIETKJHYFKZYSGMINRWTQTWOWTIEULBKZGODSARTVGXANELVOEFSITBOWGEZGZKMSGNARKVWCNQTXWRSEGXZSAWQSSGXDGDESOAGMVWKCOGQYSIADVSNSGVANGNYBKLOKZCNQNYSL


Consideremos las pistas del ayudante y las últimas 4 letras del criptograma, estas son las correspondientes a **NYSL**, veamos a que valores están mapeados, cabe recalcar que estamos usando un alfabeto de 26 letras.

In [2]:
print(ALFABETO['N'])
print(ALFABETO['Y'])
print(ALFABETO['S'])
print(ALFABETO['L'])

13
24
18
11


Consideremos estos valores en una matriz

In [3]:
matriz_cifrado = [
    [13,24],
    [18,11]
]

Por las pistas sabemos que la última letra del criptograma en el texto original era una **Z** y la penúltima era una **S** (curiosamente es la misma). Veamos a que valores están mapeados estas letras

In [4]:
print(ALFABETO['S'])
print(ALFABETO['Z'])

18
25


Por la clase de ayudantía sabemos que si la matriz del texto original tiene inversa, entonces podemos calcular la clave con la que se cifró. Así que generamos todas las posibles matrices invertibles con estos dos valores fijos.

In [5]:
matrices_con_inversa = []
for i in range(26):
    for j in range(26):
        matriz = [
            [i,j],
            [18,25]
        ]
        inversa = matriz_inversa(matriz)
        if inversa:
            matrices_con_inversa.append(matriz)
print(len(matrices_con_inversa))

312


Tenemos 312 posibles matrices invertibles. Ahora multipliquemos la inversa de estas matrices por la matriz resultante de las últimas 4 letras, el resultado de esta multiplicación será una posible clave que se utilizó para cifrar el texto original.

In [6]:
claves = []
for matriz in matrices_con_inversa:
    inversa = matriz_inversa(matriz)
    clave = matriz_modulo(multiplica_matriz_x_matriz(inversa,matriz_cifrado),26)
    claves.append(clave)

Ahora tenemos que calcular la inversa de todas las posibles claves y utilizarlas para descifrar el criptograma, alguna de estas tiene que producir un texto con sentido. Los resultados los guardamos en un archivo `.txt`

In [7]:
with open('./archivos/posibles_descifrados.txt', 'w') as file:
    for clave in claves:
        clave_inversa = matriz_inversa(clave)
        descifrado = cifrar_hill(criptograma, clave_inversa)
        file.write(descifrado+"\n")

Después de revisar los resultados, podemos observar que hay uno que tiene coherencia, este se encuentra en la linea `217` del archivo. Por lo que este debio haber sido descifrado por la clave en la posición 216, por lo que la clave para descifrar fue la siguiente:

In [8]:
inversa = matriz_inversa(claves[216])
print(inversa)
print(cifrar_hill(criptograma,inversa))

RIQUEZAFAMAPODERGOLDROGERELREYDELOSPIRATASCONQUISTOTODOLOQUEELMUNDOPUEDEOFRECERSUSULTIMASPALABRASANTESDEMORIRHICIERONQUELAGENTECORRIERAHACIAELOCANOQUIERENMITESOROSILOBUSCANPUEDENQUEDARSECONELTODOLOQUEOBTUVEDEESTEMUNDOLOENCONTRARANAHLOSHOMBRESSEDIRIGIERONALAGRANRUTAMARTIMAACUMPLIRSUSSUENOSYASICOMENZOLAGRANERADELOSPIRATASZ


Si agregamos espacios queda de la siguiente forma:

In [9]:
"""
RIQUEZA FAMA PODER GOLD ROGER EL REY DE LOS PIRATAS CONQUISTO 
TODO LO QUE EL MUNDO PUEDE OFRECER SUS ULTIMAS PALABRAS ANTES 
DE MORIR HICIERON QUE LA GENTE CORRIERA HACIA EL OCANO QUIEREN
MI TESORO SI LO BUSCAN PUEDEN QUEDARSE CON EL TODO LO QUE OBTUVE 
DE ESTE MUNDO LO ENCONTRARAN AH LOS HOMBRES SE DIRIGIERON A LA 
GRAN RUTA MARTIMA A CUMPLIR SUS SUENOS Y ASI COMENZO LA GRAN ERA 
DE LOS PIRATAS Z
"""

'\nRIQUEZA FAMA PODER GOLD ROGER EL REY DE LOS PIRATAS CONQUISTO \nTODO LO QUE EL MUNDO PUEDE OFRECER SUS ULTIMAS PALABRAS ANTES \nDE MORIR HICIERON QUE LA GENTE CORRIERA HACIA EL OCANO QUIEREN\nMI TESORO SI LO BUSCAN PUEDEN QUEDARSE CON EL TODO LO QUE OBTUVE \nDE ESTE MUNDO LO ENCONTRARAN AH LOS HOMBRES SE DIRIGIERON A LA \nGRAN RUTA MARTIMA A CUMPLIR SUS SUENOS Y ASI COMENZO LA GRAN ERA \nDE LOS PIRATAS Z\n'

Se buscó el fragmento en google y se encontró que este es una [narración del opening de one piece](https://onepiece.fandom.com/es/wiki/Narraci%C3%B3n_de_los_openings). Agregando las letras faltantes, quitando las sobrantes, con signos de puntuación y cambiando la n por ñ, queda de la siguiente forma:

<em>Riqueza, Fama, Poder. Gold Roger, el rey de los piratas conquistó
todo lo que el mundo puede ofrecer. Sus últimas palabras antes 
de morir hicieron que la gente corriera hacia el oceano. ¿Quieren
mi tesoro? si lo buscan pueden quedarse con él. Todo lo que obtuve 
de este mundo lo encontraran. Los hombres se dirigieron a la 
gran ruta marítima a cumplir sus sueños y así comenzó la gran era 
de los piratas</em>
